modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+        let mut parts: Vec<&str> = s.split('/').collect();
+        let last_part = parts.pop().expect("split() returned an empty iterator");
+
+        let (base_denom, trace_path) = {
+            if last_part == s {
+                (Denom::from_str(s)?, TracePath::default())
+            } else {
+                let base_denom = Denom::from_str(last_part)?;
+                let trace_path = TracePath::try_from(parts)?;
+                (base_denom, trace_path)
+            }
+        };
+
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl TryFrom<RawDenomTrace> for DenomTrace {
+    type Error = Error;
+
+    fn try_from(value: RawDenomTrace) -> Result<Self, Self::Error> {
+        let base_denom = Denom::from_str(&value.base_denom)?;
+        let trace_path = TracePath::from_str(&value.path)?;
+        Ok(Self {
+            trace_path,
+            base_denom,
+        })
+    }
+}
+
+impl From<DenomTrace> for RawDenomTrace {
+    fn from(value: DenomTrace) -> Self {
+        Self {
+            path: value.trace_path.to_string(),
+            base_denom: value.base_denom.to_string(),
+        }
+    }
+}
+
+impl From<Denom> for DenomTrace {
+    fn from(denom: Denom) -> Self {
+        Self {
+            trace_path: Default::default(),
+            base_denom: denom,
+        }
+    }
+}
+
+impl fmt::Display for DenomTrace {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.trace_path.0.is_empty() {
+            write!(f, "{}", self.base_denom)
+        } else {
+            write!(f, "{}/{}", self.trace_path, self.base_denom)
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, From)]
+pub struct HashedDenom(Vec<u8>);
+
+impl From<&DenomTrace> for HashedDenom {
+    fn from(value: &DenomTrace) -> Self {
+        let mut hasher = Sha256::new();
+        hasher.update(value.to_string().as_bytes());
+        let denom_bytes = hasher.finalize();
+        Self(denom_bytes.to_vec())
+    }
+}
+
+impl fmt::Display for HashedDenom {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let denom_hex = String::from_utf8(hex::encode_upper(&self.0)).map_err(|_| fmt::Error)?;
+        write!(f, "{}/{}", IBC_DENOM_PREFIX, denom_hex)
+    }
+}
+
+impl FromStr for HashedDenom {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<&str> = s.split('/').collect();
+        if parts.len() != 2 || parts[1].trim().is_empty() {
+            return Err(Error::malformed_hash_denom());
+        } else if parts[0] != IBC_DENOM_PREFIX {
+            return Err(Error::missing_denom_ibc_prefix());
+        }
+
+        let bytes = hex::decode_upper(parts[1]).map_err(Error::parse_hex)?;
+        Ok(Self(bytes))
+    }
+}
+
+/// A decimal type for representing token transfer amounts.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Display, From)]
+pub struct Amount(U256);
```

Comment:
How are users expected to use this struct? For example, how do you implement `BankKeeper::burn_coins()` when I don't have a subtract operation, and that it's not possible for me to get to the underlying `U256`?

Reply:
Fixed by implementing `Into<U256>` and exposing `checked_add/sub()`. https://github.com/informalsystems/ibc-rs/pull/1989/commits/e16bccabaf4ce32bccdcbad4bf38625fb9572825